src/compiler/checker.ts

```diff
@@ -4583,7 +4583,10 @@ namespace ts {
                 let requireOptionalProperties = relation === subtypeRelation && !(source.flags & TypeFlags.ObjectLiteral);
                 for (let targetProp of properties) {
                     let sourceProp = getPropertyOfType(source, targetProp.name);
+
                     if (sourceProp !== targetProp) {
+                        // sourceProp !== targetProp -- ie: source and target have distinct declarations with the same name

```

Comment:
It checks the type arguments first, but then if they fail (in this case they do), it checks the properties, and reaches here.